### PR TITLE
refactor(env): Change the way the environment and retina are initialized

### LINKED ISSUE #9464: [Bug]: Single or first StaticCanvas on page is blurry / doesn't respect enableRetinaScaling
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have read and followed the [Issue Tracker Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md#%EF%B8%8F-issue-tracker)
- [X] I have searched and referenced existing issues and discussions
- [X] I am filing a **BUG** report.
- [X] I have managed to reproduce the bug **after** upgrading to the **latest** version
- [X] I have created an accurate and minimal reproduction

### Version

6.0.0-beta14

### In What environments are you experiencing the problem?

Firefox, Chrome, Safari

### Node Version (if applicable)

None

### Link To Reproduction

https://jsfiddle.net/klevendev/quywt9ev

### Steps To Reproduce

1. Initialize two instances of StaticCanvas with enableRetinaScaling = true
2. Add some text to them
3. Display on a Retina screen

### Expected Behavior

I expected to see crisp text on both canvas elements.

### Actual Behavior

I saw blurry text on the first

image

---

The first StaticCanvas is always blurry on Retina. 

You can uglyhack around it by writing: 

`const foo = new StaticCanvas();`

Before your "proper" canvas. But that probably needs to be disposed and managed in some way.

### Error Message & Stack Trace

_No response_

### PR DESCRIPTION
**Pull request by ShaMan123**
## Motivation

closes #9464

## Description

StaticCanvas would call `_isRetinaScaling` before fabric would call `getEnv`.
In the browser env the devicePixelRatio config value was set during the first `getEnv` call (a contrain enforced to support SSR).
This resulted in a StaticCanvas that was initialized with a wrong devicePixelRatio value (defaulting to 1 in config)

## Changes

- rm `Canvas#_isRetinaScaling` - it did a useless check that is done in `getRetinaScaling`
- moved the node `setEnv` call the `index.node.ts` to keep the side effects in the index file.
- exposed `getDevicePixelRatio` that checks the `config.devicePixelRatio` and fallbacks to the env value. Having it evaluated each call is another fix bacause the value can change when moving the browser between screens
- `config.devicePixelRatio` has been made optional.

## TODO

I think we need to suport devicePixelRatio change event

## In Action

https://codesandbox.io/s/fabric-vanillajs-sandbox-forked-pc9hvk

### CODE REVIEW

**Comment by asturur** [comment]
Before you use a lot of time on this plan, devicePixelRatio needs to be configurable, so its place is in the config.
The rest of the cleanup could be good.

**Comment by asturur** [comment]
I also double checke that typeof window is totally fine. I was sure, but i double checked in case there was something with our code that made it unsafe in other ways.

i tried SSR with the next template from codesanbox and i set it to 3 in node, just to be sure i wasn't running the old code, and SSR works just fine.

image

**Comment by ShaMan123** [comment]
What did you set to 3? devicePixelRatio?
What use case is that?
The use case I am talking about is using the window value and that can not be done on import:

image

**Comment by ShaMan123** [comment]
> Before you use a lot of time on this plan, devicePixelRatio needs to be configurable, so its place is in the config. The rest of the cleanup could be good.

Why? Please explain. I do not understand what is the use case

File: src/env/index.ts [hunk 1]
```diff
@@ -30,9 +30,18 @@ export const setEnv = (value: TFabricEnv) => {
...
 export const getFabricDocument = (): Document => getEnv().document;
 
 export const getFabricWindow = (): (Window & typeof globalThis) | DOMWindow =>
   getEnv().window;
+
+/**
+ * @returns the config value if defined, fallbacks to the environment value
+ */
+export const getDevicePixelRatio = () =>
+  Math.max(config.devicePixelRatio ?? getEnv().devicePixelRatio, 1);
```
Comment (ShaMan123) [inline src/env/index.ts:47]: I do not like this but it is not a big deal if you think dpr belongs to config

**Review by ShaMan123 — COMMENTED** [review]
updated description
this PR is ready

**Comment by asturur** [comment]
> What did you set to 3? devicePixelRatio? What use case is that? The use case I am talking about is using the window value and that can not be done on import:
> 
> image

Yeah this you do will error. if you put a typeof in front of window and check for it to be not undefined is good.
you can't access window, you can check if is defined and the access it.

In my comment i specified i used 3, because i wanted to be sure i wasn't running the old code that was returning 1 on node, so in the config i added:
`typeof window !== 'undefined' ? window.devicePixelRatio : 3` just to be sure i was running my fresh built code and not some stale old one i had somewhere in the file system.

devicePixelRatio was configurable by choice, because sometimes you may want to limit it ( in some devices is 3.5 or 4) or simply you want to set it to 2 and you want to ignore extra value that is there because of the user zooming the window.

**Comment by ShaMan123** [comment]
> you want to ignore extra value that is there because of the user zooming the window.

How does zooming affect dpr?

> Yeah this you do will error. if you put a typeof in front of window and check for it to be not undefined is good.
you can't access window, you can check if is defined and the access it.

This is my point exactly. You can't do that if you want to reliably access a window value in SSR because there will be no window at the time of the call. And then it means that fabric will use 1 as the dpr value instead of the window dpr value.

**Review by ShaMan123 — COMMENTED** [review]
@asturur I applied a change to align better to what you suggested
Please review it again

**Review by asturur — APPROVED** [review]

**Pull request merged**